mathics/packages/DiscreteMath/CombinatoricaV2.0.m

```diff
@@ -0,0 +1,3422 @@
+
+TopologicalSort::usage = "TopologicalSort[g] returns a permutation of the vertices of directed acyclic graph g such that an edge (i,j) implies vertex i appears before vertex j."
+
+TransitiveClosure::usage = "TransitiveClosure[g] finds the transitive closure of graph g, the superset of g which contains edge {x,y} iff there is a path from x to y."
+
+TransitiveQ::usage = "TransitiveQ[g] returns True if graph g defines a transitive relation."
+
+TransitiveReduction::usage = "TransitiveReduction[g] finds the smallest graph which has the same transitive closure as g."
+
+TranslateVertices::usage = "TranslateVertices[v,{x,y}] adds the vector {x,y} to each vertex in list v."
+
+TransposePartition::usage = "TransposePartition[p] reflects a partition p of k parts along the main diagonal, creating a partition with maximum part k."
+
+TransposeTableau::usage = "TransposeTableau[t] reflects a Young tableau t along the main diagonal, creating a different tableau."
+
+TravelingSalesmanBounds::usage = "TravelingSalesmanBounds[g] computes upper and lower bounds on the minimum cost traveling salesman tour of graph g."
+
+TravelingSalesman::usage = "TravelingSalesman[g] finds the optimal traveling salesman tour in graph g."
+
+TreeQ::usage = "TreeQ[g] returns True if graph g is a tree."
+
+TriangleInequalityQ::usage = "TriangleInequalityQ[g] returns True if the weight function defined by the adjacency matrix of graph g satisfies the triangle inequality."
+
+Turan::usage = "Turan[n,p] constructs the Turan graph, the extremal graph on n vertices which does not contain K[p]."
+
+TwoColoring::usage = "TwoColoring[g] finds a two-coloring of graph g if g is bipartite."
+
+UndirectedQ::usage = "UndirectedQ[g] returns True if graph g is undirected."
+
+UnionSet::usage = "UnionSet[a,b,s] merges the sets containing a and b in union-find data structure s."
+
+UnweightedQ::usage = "UnweightedQ[g] returns True if all entries in the adjacency matrix of graph g are zero or one."
+
+V::usage = "V[g] gives the order or number of vertices of graph g."
+
+VertexColoring::usage = "VertexColoring[g] uses Brelaz's heuristic to find a good, but not necessarily minimal, vertex coloring of graph g."
+
+VertexConnectivity::usage = "VertexConnectivity[g] computes the minimum number of vertices whose deletion from graph g disconnects it."
+
+VertexCoverQ::usage = "VertexCoverQ[g,c] returns True if the vertices in list c define a vertex cover of graph g."
+
+Vertices::usage = "Vertices[g] returns the embedding of graph g."
+
+WeaklyConnectedComponents::usage = "WeaklyConnectedComponents[g] returns the weakly connected components of directed graph g."
+
+Wheel::usage = "Wheel[n] constructs a wheel on n vertices, which is the join of K[1] and Cycle[n-1]."
+
+WriteGraph::usage = "WriteGraph[g,f] writes graph g to file f using an edge list representation."
+
+Begin["`private`"]
+PermutationQ[p_List] := (Sort[p] == Range[Length[p]])
```

Comment:
@rocky, this is the line that I was talking about.

Reply:
> @rocky, this is the line that I was talking about.

Then I think there is a bug in Mathics not CombinatoricaV2.0.m. I checked with the published book, and indeed appears in section 2.1 on page 55. 

When you run this on Mathematica, do you get the same behavior as Mathics?

Reply:
> > @rocky, this is the line that I was talking about.
> 
> Then I think there is a bug in Mathics not CombinatoricaV2.0.m. I checked with the published book, and indeed that code appears in section 2.1 on page 55.
> 
> When you run this on Mathematica, do you get the same behavior as Mathics?

The behaviour is the same for this definition (I changed the name to avoid a conflict with the builtin symbol):

In WMA:
```
In[1]:= MyPermutationQ[p_List]:=(Sort[p] == Range[Length[p]])                   

In[2]:= Permutations[Range[3]]                                                  

Out[2]= {{1, 2, 3}, {1, 3, 2}, {2, 1, 3}, {2, 3, 1}, {3, 1, 2}, {3, 2, 1}}

In[3]:= ptst=Permutations[Range[3]]                                             

Out[3]= {{1, 2, 3}, {1, 3, 2}, {2, 1, 3}, {2, 3, 1}, {3, 1, 2}, {3, 2, 1}}

In[4]:= MyPermutationQ[ptst]                                                    

Out[4]= {{1, 2, 3}, {1, 3, 2}, {2, 1, 3}, {2, 3, 1}, {3, 1, 2}, {3, 2, 1}} == 
 
>    {1, 2, 3, 4, 5, 6}

```


In Mathics:

```
In[1]:= MyPermutationQ[p_List]:=(Sort[p] == Range[Length[p]])
Out[1]= None

In[2]:= Permutations[Range[3]]
Out[2]= {{1, 2, 3}, {1, 3, 2}, {2, 1, 3}, {2, 3, 1}, {3, 1, 2}, {3, 2, 1}}

In[3]:= ptst=Permutations[Range[3]]                                             
Out[3]= {{1, 2, 3}, {1, 3, 2}, {2, 1, 3}, {2, 3, 1}, {3, 1, 2}, {3, 2, 1}}

In[4]:= MyPermutationQ[ptst]
Out[4]= {{1, 2, 3}, {1, 3, 2}, {2, 1, 3}, {2, 3, 1}, {3, 1, 2}, {3, 2, 1}} == {1, 2, 3, 4, 5, 6}
```


In both cases, the result is not the expected (`True`)

Reply:
Ok. Then I will contact the authors about this, and we can change what is in this branch adding a comment. 

The book says this was uses Mathematica 4.2. Apparently things have changed since then.

Reply:
Or maybe I just do not understand what they were trying to do with this. The behavior of the involved symbols didn't change too much since v4.